### ISSUE TITLE
Long matches are indented incorrectly

### ISSUE THREAD

**Issue by cmeeren** [fsprojects/fantomas#1129]
Issue created from [fantomas-online](https://fsprojects.github.io/fantomas-tools/#/fantomas/preview?data=N4KABGBEDGD2AmBTSAuKBbAhgF2gCzCwA8BJdAcwHUBLebPAGjESO0QDsBna2dsAd2r0AOgCdh7AD5gAyrHSJCmUhRp1Gs%2BYoAU24ZABWAB3L6w0-ccSnI5sPqPsbd-eWoAzMxcjZymL-aQAEboRvoAlGCYnMyskQC0AHxg7rCwYhLSAPpgSSlpGeyQDOBQnIjY2NROnKhgANqlEKAQrVAAJNgAnkbIaJDV2MVNrZDtAG6YADYArn0NI20ADCVtbZAk7Ejs2DLUAF7Iq2sQACyLYAC6IwC%2Bx80XY929dQM7wydjk7PzjScQAEZ7mtIABZZQAGWqiAhHHI9A%2B-wBACYlhdrm07iMWiDOj15sE0lNEesJtM5nU-v9kcD1jJEOhqHAprwAILYACiWwA8u4oewjhcIO5puV0bdgTjSc8CUEiSTRmSfpShWAAMy00YyIyYaCIABCiFSokQAAVMKJMAo2KIFW1sKI5uLMZLHniXv05bBiZqvuTfqrTpqIJBtbqDUbYCaIbB%2BIhRNBoohNuNYImqrw7a0HU6ThjWli2lLFTLXl6fW7vhS0FSTgBWYNQMN6w3GxAAVSMvQTSZTaZwPCKjZFUzFeYl2Ldpc98t9SurC3%2BYAAbI3QzqW5GTQBhKbRTjb3icHPQbBRrPC0WIZ0F12fd2y2eV-0qpcAdjXzYjbdBDKC8YvFIrxvCBC1aYsQwfMsn3vKsAyXAAOT8N2-KNEEPFkh1VEcxzWfNQLvXFpygcssz9ZUa1VABOZDw1ZdwbUPdAsEAnNr3HF1J3vYjCW9Mj53gpEVlVddw1bND6UZZlM2HYCONvLiiPxaC%2BLnODXyRIERK-eibUkplvRk1U2JAsAwIebjlJnVTnwoxckRpETNm2bBuXYAAVUQuhoBFZNHdi8InIspyskiYNxdTKKXAENW0lDWVEWAZi2AARRApmoRkbVYx0ArafCzMI6VQreIY1JfKKkSDETwVIdx3LwDgOX8mQ8CjbA1F8wM0XkgjFOKj0oEGfjIvsk4AQbGrlE2dxqCIbluxwKMOSIIwTU4bgjKXOsesCzjgsswbSpGiqxrWAFVymogACVEDgUR4E6vBANOXb8qC8CQqO4byrs2tzo-K6EstLpuVEKFjyel63taArzLACCOh4n7bIXf62gBJCroANX9fVqngaoqFoLql1e0z4cRp4SpR2DTvR1oARoq6ADFktPQd8a2ImocbcnesK-qSxK0jfrR1VUTXUEZimKoMoFfUWWgABrfVLRVipODcmQrXQ70ZnQLClxwvLYY%2BiylKO0XUcEk5kS0pdIAAOUQfh5YjbA4w4dz8TS2b2CEQdWS2X90H-URaj83D3v2z7DsfGy6b%2BiXHMdgBpRBECMEh6sa9hNh1hR%2BUFY25L2hSDstglaYi%2BmJdix3auaxlODwXnuoporhat8LSVGhmIGRarHb2JwpkQBLyANjgOsQIJQQQEv-hNzuhcgnjraT8Wl2RSbHdZDLyHYNn2A53g9iPnAZhNdzYGcmeBy25ey5jiu46rlSKy3221mRS799lvGdgA5xgwl4OQUO4c-bVEDkeQCK8BaUy%2BgnL%2Btdk470BiPB0TJsALyQPAl%2BZtMSlEuMcAYnAWbcDqCvG4QA)

#### Code

```fsharp
match maxImgWidth, extension with
| Some maxImgWidth, Some (("jpg" | "jpeg" | "png" | "gif" | "tga" | "bmp") as ext) -> foo
| _ -> foo

```
            

#### Result

```fsharp
match maxImgWidth, extension with
| Some maxImgWidth,
  Some (("jpg"
  | "jpeg"
  | "png"
  | "gif"
  | "tga"
  | "bmp") as ext) -> foo
| _ -> foo
```
            
#### Problem description

IMHO it doesn't look good with this combination of parens, indentation and line breaks. I'm open to suggestions, e.g.

```fsharp
match maxImgWidth, extension with
| Some maxImgWidth,
  Some ((
    "jpg"
    | "jpeg"
    | "png"
    | "gif"
    | "tga"
    | "bmp" ) as ext) -> foo
| _ -> foo
```

```fsharp
match maxImgWidth, extension with
| Some maxImgWidth,
  Some (
     ( "jpg"
     | "jpeg"
     | "png"
     | "gif"
     | "tga"
     | "bmp") as ext) -> foo
| _ -> foo
```

#### Extra information

- [ ] The formatted result breaks by code.
- [ ] The formatted result gives compiler warnings.
- [ ] I or my company would be willing to help fix this. *(I'll help in any way I can except implementation; I don't have the capacity to ramp up on a new codebase now)*

#### Options

Fantomas Master at 09/09/2020 21:40:22 - 0d320054c533698c732b12dcf524b8545369d67d

Default Fantomas configuration

**Comment by nojaf** [comment]
Hey @cmeeren, at first glance I think we can easily have something like this:
```fsharp
match maxImgWidth, extension with
| Some maxImgWidth,
  Some (("jpg"
         | "jpeg"
         | "png"
         | "gif"
         | "tga"
         | "bmp") as ext) -> foo
| _ -> foo
```

but I would need to check as `SynPat.Or` is used in other places as well.

**Comment by cmeeren** [comment]
Great! Is there a setting that controls whether all the strings in the example above appear on one line or are chopped? I tried setting all max widths to `120`, but the result is the same. IMHO they don't need to be chopped in the example above, e.g.:

```f#
match maxImgWidth, extension with
| Some maxImgWidth,
  Some (("jpg" | "jpeg" | "png" | "gif" | "tga" | "bmp") as ext) -> foo
| _ -> foo
```

**Comment by nojaf** [comment]
This might be a bit trickier than I thought.
So the AST shape of the strings is SynPat.Or which is the same shape you would get for in for example:
```fsharp
match foo with
| Bar _
| Fooey _
| Abc _ ->
    printfn "meh"
| _ ->
    printfn "no meh"
```

Here `Bar`, `Fooey` and `Abc` are also wrapped in a `SynPat.Or` construct. Meaning, putting the string scenario on one line would also put `| Bar _ | Fooey _ | Abc _ ->` on one line.

So it would be better to indicate somehow that we are in a different scenario. This might be possible to keep track of in ASTContext.

So long story short, probably doable, though we need to be careful not to break other stuff.

**Comment by cmeeren** [comment]
> Meaning, putting the string scenario on one line would also put `| Bar _ | Fooey _ | Abc _ ->` on one line.

I have no problem with that. For sufficiently short/few cases, I often format my code like that anyway.

**Comment by nojaf** [comment]
Old issue. The result today no longer reflects the problem case.
